### PR TITLE
Inactivity fix standalone

### PR DESCRIPTION
**Pull request by IMbackK**
Fixes inactivity problems on droid 4 (timeout takes way to long)

### CODE REVIEW

**Comment by IMbackK** [comment]
fixes https://github.com/maemo-leste/bugtracker/issues/428

File: event-input.h [hunk 1]
```diff
@@ -162,6 +162,46 @@ static const int *const touch_event_keys[]= {
...
+	KEY_K,
+	KEY_L,
+	KEY_Z,
+	KEY_X,
+	KEY_C,
+	KEY_V,
+	KEY_B,
+	KEY_N,
+	KEY_M,
+	-1
```
Comment (freemangordon) [inline event-input.h:198]: Did you get that list from some documentation, that states like "if you have those keys, then it is a keyboard" or you made it up? Also, this will not match power button on n900, for example, so once we remove match by name (and we'll have to soon or later), we'll have lots of things broken. Please rethink that criteria.

Comment (freemangordon) [inline event-input.h:198]: Another issue I see here - didn't try it, but most-probably we'll end up with keyboard hook attached twice on n900 for the HW kbd - once because of the existing logic and once more - because of that new logic. That shall be fixed in the same commit.

File: devlock.c [hunk 2]
```diff
@@ -91,6 +91,8 @@ static gboolean device_autolock_timeout_cb(gpointer data)
 	(void)data;
```
Comment (freemangordon) [inline devlock.c]: Please split that into separate commits:<br>1.the one with debug prints<br>2.functional changes commit.<br>Also, please elaborate (with details) in the commit message what is the problem and how it is solved. Why it works on n900 and not on d4?

File: debian/control [hunk 3]
```diff
@@ -24,7 +24,6 @@ Build-Depends:
  libcal-dev (>= 0.2.4),
  libdevlock-dev,
  libupower-glib-dev (>=1:0.99.7.5),
- libsystemd-dev,
```
Comment (freemangordon) [inline debian/control:27]: Who added that in the first place?!? ACK, but why is this commit part of that PR?

File: event-input.c [hunk 4]
```diff
@@ -529,6 +529,9 @@ static void match_and_register_io_monitor(const gchar *filename)
 		/* If the driver for the event file is blacklisted, skip it */
 		close(fd);
 		goto EXIT;
+	} else if (strstr(filename, "event") == NULL) {
```
Comment (freemangordon) [inline event-input.c:538]: ACK, but I guess that shall go in a separate PR.

**Review by freemangordon — CHANGES_REQUESTED** [review]

File: event-input.h (hunk 1 above)
Comment (IMbackK) [reply event-input.h:198]: This is quite similar to how the kernel defines a keyboard, it defines  a kyboard as anything reporting keycodes <255. 
duplicating this behaviour would be tedious with the current code so i setteled on this.

this should explicitly NOT match devices with power keys that is handled by the next match down: power_event_types

Comment (IMbackK) [reply event-input.h:198]: No there is no way the code ever matches a device twice, every match is guarded by an else if

Comment (IMbackK) [reply event-input.h:198]: the old code comes first so it takes precedence

File: devlock.c (hunk 2 above)
Comment (IMbackK) [reply devlock.c]: 1. Id rather not split it.

2. the current code is supposed to work like this: a event reaches https://github.com/maemo-leste/mce/blob/3cbee89d364eea0211fa7a37493e39a6c2b22192/event-input.c#L168 here the event gets processed. The callback then uses a timer to suspend the iomonitor for 1 second: https://github.com/maemo-leste/mce/blob/3cbee89d364eea0211fa7a37493e39a6c2b22192/event-input.c#L200 during this time n events pile up in the evdev device. When the timer times out it will re-enable the iomonitor and mce_resume_io_monitor will seek to the end of the file, effectively skiping the n events that piled up.

3. This, however, is broken because you can not seek evdev devices so the seek fails. Upon resume of the iomonitor touchscreen_cb will instead simply get the next event that piled up during the 1 second suspend and will then suspend the iomonitor again. this will repeat until every event is processed. Because the droid4's touchscreen creates events at a rate of 400Hz while swiping, this will take a _very_ _very_ long time. During this time mce will think the device is being interacted with so the device will enter device_idle state only after a very long time.

4. this pr solves this problem by supplying the io_monitor with the function mcs_io_monitor_seek_to_end. This function attempts to seek to the end of a file. If it can not it will instead read() to the end of the file and discard the data.

Comment (IMbackK) [reply devlock.c]: regarding why this works on n900. I dont know and cant confirm that it dose. sicelo on irc claims it works. assumeing it dose:

1. maybe the kernel allows seeking in some instances (why?)
2. maybe this never worked but the update rate of the n900 touchscreen is low enough that this is not so appearant.

File: debian/control (hunk 3 above)
Comment (IMbackK) [reply debian/control:27]: afaik pali added this. He added some systemd code to mce, not sure what it dose. It is however behind a ifdef and the packaging scripts do not set this define. therefore this dependancy is invalid.

I thought it a to insignificant change to push as its own pr, but i dont care, it can be one.

File: event-input.c (hunk 4 above)
Comment (IMbackK) [reply event-input.c:538]: no this is a vital part of this pr. On the droid4 (and likely others)  there are legacy mouse* devices available. MCE opens these devices as evdev devices, and thus gets invalid data. This causes the device idle watcher to misbehave and time out to late.

Comment (freemangordon) [reply event-input.c:538]: ok

File: debian/control (hunk 3 above)
Comment (freemangordon) [reply debian/control:27]: ok, lets leave it as it is

File: event-input.h (hunk 1 above)
Comment (freemangordon) [reply event-input.h:198]: ok

File: mce-io.c [hunk 5]
```diff
@@ -868,6 +856,39 @@ void mce_unregister_io_monitor(gconstpointer io_monitor)
...
+	
+	if(!seek_success) {
+		int bytes_read;
+		char buffer[1024];
+		int count = 0;
+		do {
+			g_io_channel_read_chars(iomon->iochan, buffer,
+						1024,
+						&bytes_read, &error);
+			count++;
```
Comment (freemangordon) [inline mce-io.c]: This variable is not used, unless I am missing something.

File: devlock.c (hunk 2 above)
Comment (freemangordon) [reply devlock.c]: What is the error seek fails with?

Comment (freemangordon) [reply devlock.c]: I had written a long answer etc., however, github somehow managed to lose it :( . Anyway, in short it was - lets wait until me or @MerlijnWajer is able to test on n900 (and maybe pinephone) to see if seek works there or not. If it works, I think we shall understand why (and why it doesn't on d4).

File: mce-io.c [hunk 6]
```diff
@@ -868,6 +856,39 @@ void mce_unregister_io_monitor(gconstpointer io_monitor)
...
+	
+	error = NULL;
+	
+	if(!seek_success) {
+		int bytes_read;
+		char buffer[1024];
+		int count = 0;
+		do {
+			g_io_channel_read_chars(iomon->iochan, buffer,
+						1024,
```
Comment (freemangordon) [inline mce-io.c]: sizeof(buffer)?

File: mce-io.c (hunk 5 above)
Comment (IMbackK) [reply mce-io.c]: upps yes thats left over form debugging

File: mce-io.c (hunk 6 above)
Comment (IMbackK) [reply mce-io.c]: sounds good

File: debian/control (hunk 3 above)
Comment (MerlijnWajer) [reply debian/control:27]: I think Spinal added it.

File: mce-io.c [hunk 7]
```diff
@@ -450,23 +458,10 @@ static gboolean io_error_cb(GIOChannel *source,
 			"called with iomon == NULL!");
 		goto EXIT;
 	}
-
-	switch (iomon->error_policy) {
-	case MCE_IO_ERROR_POLICY_EXIT:
+
```
Comment (MerlijnWajer) [inline mce-io.c]: whitespace - maybe run indent -linux with the right options on this to avoid this?

**Comment by IMbackK** [comment]
While this pr fixes event_input.c behavior and iniactivity.c device_inactive_pipe = true gets set by inactivity.c 100% reliably inactivity locking remains extremely broken.

This starts off with the fact that inactiviy.c setting device_inactive_pipe = true dose absolutely nothing at all. Only devlock.c acts on device_inactive_pipe = true and it acts in the same way as it dose with device_inactive_pipe = false and dose not result in the display turning off.  This can be easly proven by unloading this module, device behavor is the same while it would be expected that the device never times out and disables the display due to a lack of user input.

Instead the device idleness timers that acctualy DO work appear to be in systemui but the implentation is extreamly broken:
1. sometimes it never turns off the display while the keyboard is open
2. sometimes it turns off the display way to late
3. sometimes it locks but due to a race with mce the device ends in a locked state with the display on
4. sometimes it is out of sync with mce and the timer in tklock.c that is supposed to turn of the display while tklock is shown, even though tklock is not currently enabled. Which causes the device to lock way to early.

MCE dose enable the display on device_inactive_pipe = false, and the only reason why the device now behaves correctly more often with this pr is because with out it mce was infact enabling the display every second, masking systemui/tklock.c's efforts to turn the display off

**Comment by IMbackK** [comment]
```
21:11:36 <uvos> so regarding https://github.com/maemo-leste/mce/pull/12
21:11:38 <Entitlement> uvos - [ Inactivity fix standalone by IMbackK · Pull Request #12 · maemo-leste/mce · GitH... ]
21:11:56 <uvos> you said it what this fixes is also broken in fremantle
21:12:04 <uvos> at least it seams so
21:12:08 <freemangordon> not really sure what's going on
21:12:17 <freemangordon> it needs more investigation
21:12:34 <uvos> ok but you understand what i observe on d4
21:12:38 <freemangordon> on fremantle most of the time it works
21:12:44 <uvos> also the evdev can not be seeked
21:12:46 <freemangordon> I think yes
21:13:01 <freemangordon> I see the same on the VM
21:13:05 <uvos> well it spits out EPIPE
21:13:13 <freemangordon> but it could be because there is no TS there
21:13:18 <uvos> thats the kernel telling us we cant seek
21:13:29 <uvos> freemangordon: the bug also manifests on other input devices
21:13:41 <uvos> freemangordon: its just less pronounced because the event rate is lower
21:13:56 <freemangordon> yes, I understand that, but I want to dig a little bit more
21:14:01 <uvos> freemangordon: all right
21:14:08 <freemangordon> to have a complete understanding why it works sometines
21:14:15 <freemangordon> *sometimes
21:14:15 <uvos> ok
21:14:20 <uvos> afaik it never works
21:14:21 <uvos> but ok
```

**Comment by IMbackK** [comment]
Regardless of why this may or may not work on fremantle, I would also like to note that having a fallback for unseekable devices, as this pr provides, in io-monitor is helpful in any case, as mce might need to monitor ex. a pipe in the future.

File: mce-io.c [hunk 8]
```diff
@@ -370,11 +369,9 @@ static gboolean io_chunk_cb(GIOChannel *source,
 
 	chunk = g_malloc(iomon->chunk_size);
 
-	do {
```
Comment (freemangordon) [inline mce-io.c]: why is that loop removed?

Comment (IMbackK) [reply mce-io.c]: this loop can cause a deadlock. Evdev uses EAGAIN to signify that no event is currently available, this state persists until the next event is triggerd on the event device. If for any reason the mce code becomes unaligned (in respect to sizeof(event) blocks) with the event file, such as when the input layer is not done writing to the file, mce will get stuck in an endless loop until the user presses another key.

**Comment by freemangordon** [comment]
I am fine with the latest version, @MerlijnWajer - shall we merge it?
